lib/charms/grafana_agent/v0/cos_machine.py

```diff
@@ -0,0 +1,257 @@
+        for event in self._refresh_events:
+            self.framework.observe(event, self.update_relation_data)
+
+    def update_relation_data(self, event):
+        """Trigger the class to update relation data."""
+        if isinstance(event, RelationEvent):
+            relations = [event.relation]
+        else:
+            relations = self._charm.model.relations[self._relation_name]
+
+        for relation in relations:
+            relation.data[self._charm.app].update({"config": self._generate_updated_data()})
+
+    def _generate_updated_data(self) -> str:
+        """Collate the data for each nested databag and return it."""
+        data = {
+            "metrics": {
+                "scrape_jobs": self._scrape_jobs,
+                "alert_rules": self._metrics_alert_rules,
+            },
+            "logs": {
+                "targets": self._logs_slots,
+                "alert_rules": self._log_alert_rules,
+            },
+            # Probably doesn't actually need to be nested
+            "dashboards": {
+                "dashboards": self._dashboards,
+            },
+        }
+
+        return json.dumps(data)
+
+    @property
+    def _scrape_jobs(self) -> List[Dict]:
+        """Return a prometheus_scrape-like data structure for jobs."""
+        job_name_prefix = self._charm.app.name
+        return [
+            {"job_name": f"{job_name_prefix}_{key}", **endpoint}
+            for key, endpoint in enumerate(self._metrics_endpoints)
+        ]
+
+    @property
+    def _metrics_alert_rules(self) -> Dict:
+        """Use (for now) the prometheus_scrape AlertRules to initialize this."""
+        alert_rules = MetricsAlerts()
+        alert_rules.add_path(self._metrics_rules, recursive=self._recursive)
+        return alert_rules.as_dict()
+
+    @property
+    def _log_alert_rules(self) -> Dict:
+        """Use (for now) the loki_push_api AlertRules to initialize this."""
+        alert_rules = LogAlerts()
+        alert_rules.add_path(self._logs_rules, recursive=self._recursive)
+        return alert_rules.as_dict()
+
+    @property
+    def _dashboards(self) -> List[str]:
+        dashboards = []
+        for d in self._dashboard_dirs:
+            for path in Path(d).glob("*"):
+                dashboards.append(self._encode_dashboard_content(path.read_bytes()))
+
+        return dashboards
+
+    def _encode_dashboard_content(self, content: Union[str, bytes]) -> str:
+        if isinstance(content, str):
+            content = bytes(content, "utf-8")
+
+        return base64.b64encode(lzma.compress(content)).decode("utf-8")
+
+
+class COSMachineDataChanged(EventBase):
+    """Event emitted when a `COSMachineProvider` joins or updates data."""
+
+
+class COSMachineConsumerEvents(ObjectEvents):
+    """Event descriptor for events raised by `COSMachineConsumer`."""
+
+    data_changed = EventSource(COSMachineDataChanged)
+
+
+class COSMachineConsumer(Object):
+    """Provider class for the cos_machine interface."""
```

Comment:
```suggestion
    """Integration endpoint wrapper for the Requirer side of the cos_machine interface."""
```

Also, why isn't it `COSMachineRequirer`?

Reply:
You are right, should be Requirer.